### ISSUE TITLE
Memory arena for boxes and beyond

### ISSUE THREAD

**Issue by dginev** [dginev/latexml-oxide#8]
This is a performance issue where the Rust implementation has a first chance of showing its potential for optimization.

To set things up, I created an artificial performance test, by redundantly copying the `tokenize/equality` test onto itself, for a total of 10,000 lines. Now that `rtx` can pass the test, it allows for an authentic performance comparison.

The initial times were:

|tool |time (s) |
|:-:|-:|
| pdflatex | 0.3 |
| rtx | 2.2 |
| latexml | 21.0 |

Next, I used `valgrind` and `kcachegrind` to look for bottlenecks. Unsurprisingly, most of the time was spend allocating and deallocating memory for data structures, and in particular clone calls and vector allocations. 

To solve, first I used the obvious techniques in the high profile cases from the performance profile [changeset here](https://github.com/dginev/rtx/compare/fae9bd714cec3b876c3138dc79bd95ef18cfb728...master):
 - `Cow<T>` generic wrapper for allowing to borrow in complex data flows
 - reorganize data flows where possible to avoid the need to clone due to an early move of a value 
 - use `Rc<T>` for stateful bookkeeping, to avoid frequent deeper clones

This lead to a two fold speedup for this test for `rtx` - from 2.2 s to 1.1. 

That said, the benchmark goal is pdflatex's 0.3 s, which suggests that pdflatex is already well optimized in terms of data management and does not perform any wasteful allocations - as could be expected.

The value of Rust is very clear here - any wasteful data management in the code base is marked explicitly with a `.clone` and can easily be searched for, and refactored. That said, a lot of the global Perl techniques become increasingly difficult to refactor away, and need potentially far-reaching changes to the programming model. A particularly nasty one is:

```perl
local $LaTeXML::BOX = $box;
```

which is impossible to reproduce in Rust. The heart of the problem is that the Perl model keeps a (mutable) reference to `$box` in this global variable (which is also subtly special in visibility thanks to Perl's `local` mechanics), *during* the absorbtion of said $box. In rust, the moment the absorbtion begins, you expect that either a mutable reference, or the value itself, is consumed by the absorbtion flow, hence you can not simultaneously hold/use any reference to `$box` separately from the main flow. It does not even help to use interior mutability and hold a `Rc<RefCell<T>>` of the box, as the runtime error will be triggered all the same, as there will be multiple references in the scope of a mutable one.

This is a strong indication that the technique itself is a code smell, and I tend to agree with the Rust compiler - the `$LaTeXML::Box` mechanics are hard to reason about in Perl, and they are also hard to predict and anticipate. 

I am working on refactoring this flow, and based on the patterns of box use during construction (and their bookkeeping in `Document::node_boxes`) it feels like each box should end up owned by the main document object, and passed in to the various calls by reference -- avoiding any spurious memory overuse. Mutability continues to be a point of concern, but it may be possible to isolate the issue... Will keep thinking.

**Comment by dginev** [comment]
Actually... There is *a* way with interior mutability, though unclear if it's not a poison pill to swallow, as it further dirties the Rust-purity of the project. Idea being:

- Keep a `document.box_to_absorb` equivalent of `$LaTeXML::Box`, which is typed as `Option<Rc<Digested>`, so we have a cheap ability to  clone references to the current box.
- That implies the absorbtion API has to work on immutable references to `&Digested`. That's easy enough for Tbox, but not so for `Whatsit` which can be mutated (e.g. properties set).
- Hence, we have to change `Digested` to hold a `RefCell<Whatsit>` instead, and define the `BoxOps` layer on the RefCell level. So `set_property` would be callable on immutable references and the checks for the borrowing rules will be performed at runtime. There is a small associated runtime penalty to that, but it's much cheaper than `.clone` on every Digested.

That said, we lose compiler guarantees for the whatsits are managed, and generally go closer to the "chaos" of Perl globals. So I have to think a bit if it's the only reasonable way, before swallowing the pill.

Edit: After playing around, this approach adds further issues - you can no longer reliably borrow from within a Whatsit, methods such as `get_font` and `get_property` can not obtain a value that lives long enough, due to the RefCell mechanics.

**Comment by dginev** [comment]
One thing I feel happy with is the alternative to having a `_box` attribute in each Node, namely via a dedicated field in Document:

```rust
pub node_boxes: HashMap<usize, Rc<Digested>>, // used to be _box attribute
```

the `Rc<Digested>` definitely feels adequate there.

**Comment by dginev** [comment]
Trying to `RefCell<Whatsit>` approach didn't seem compelling at all... Very far-reaching changes are required in that direction. Passing in the current box as an argument to all construction methods, or alternatively, collecting all nodes associated with the current absorbtion and swimming those up the results chain, feels much easier to manage. A few places in the perl bindings setup $LaTeXML::Box directly however, so that also needs to be considered...

For the example in question, completely disabling the `box_to_absorb` code flow improves from 1.1 to 0.9 s in runtime, or a 10-20% speedup, so maybe something to postpone solving if too tricky...

**Comment by dginev** [comment]
I found a surprising solution here, after taking all the wrong turns in the maze.

The mutability requirement for whatsits was the big stumbling block that prevented using simple Rc<Digested> wrappers. I am going to add a follow-up commit now that drops the RefCells even, just rejoicing that the solution worked.

The idea being that during absorbtion, the whatsits are provided as reference material to examine, rather than to modify - modifying their properties is of primary importance during the earlier "token invocation" stage, but not so during absorbtion. Or so I hope.

In any case, we have an escape hatch even if that assumption fails to hold - there is still mutable access ot the State, and all whatsit-property modification during absorbtion (if any), that has to be ported over from the existing bindings, can be retooled as modifying the State accordingly.

**Comment by dginev** [comment]
Solving this situation by restricting mutability - which makes the flow additionally disciplined - is actually very satisfying. And light in resource use.

**Comment by dginev** [comment]
Done: d607486c4af450e6db5d31085c04d2c827ed35e3

The large test is now 10-15% faster with this issue fixed alone, down to 0.9-1.0 seconds
